File: tests/Majako.Collections.RadixTree.Tests/Profiling.cs
using System.Diagnostics;
using Xunit.Abstractions;

namespace Majako.Collections.RadixTree.Tests
{
    public class Profiling
    {
        private readonly ITestOutputHelper _output;
        private readonly ITree<int> _sut = new ConcurrentTrie<int>();

        public Profiling(ITestOutputHelper output)
        {
            _output = output;
        }

        private void ProfileAction(Action action)
        {
            var sw = new Stopwatch();
            var memory = GC.GetTotalMemory(true) >> 20;
            sw.Start();

            action.Invoke();

            sw.Stop();
            var delta = (GC.GetTotalMemory(true) >> 20) - memory;
            _output.WriteLine("Elapsed time: {0:F}s", sw.ElapsedMilliseconds / 1000.0);
            _output.WriteLine("Memory usage: {0:F}Mb", delta);
        }

        [Fact]
        public void Profile()
        {
            ProfileAction(() =>
            {
                for (var i = 0; i < 1000000; i++)
                    _sut.Add(Guid.NewGuid().ToString(), 0);
            });
        }

        [Fact]
        public void ProfilePrune()
        {
            // insert
            for (var i = 0; i < 10000; i++)
                _sut.Add(Guid.NewGuid().ToString(), 0);

            ProfileAction(() =>
            {
                Parallel.For(0, 1000, new ParallelOptions { MaxDegreeOfParallelism = 8 }, j =>
                {
                    for (var i = 0; i < 20; i++)
                    {
                        // insert
                        _sut.Add(Guid.NewGuid().ToString(), 0);

                        // remove by prefix
                        _sut.Prune(Guid.NewGuid().ToString()[..5], out _);
                    }
                });
            });
        }
    }
}

File: tests/Majako.Collections.RadixTree.Tests/Profiling.fs
module Majako.Collections.RadixTree.Tests.Profiling

open System.Diagnostics
open Xunit
open Majako.Collections.RadixTree.Concurrent
open System
open Xunit.Abstractions

[<Literal>]
let SKIP: string = "Profiling disabled" // set to null to enable profiling
// let SKIP: string = null // set to null to enable profiling

type Scenarios(output: ITestOutputHelper) =
    let profile f =
        let sw = Stopwatch()
        let memory = GC.GetTotalMemory(true) >>> 20
        sw.Start()
        f ()
        sw.Stop()
        let delta = (GC.GetTotalMemory(true) >>> 20) - memory
        output.WriteLine(sprintf "Elapsed time: %.2f s" (float sw.ElapsedMilliseconds / 1000.0))
        output.WriteLine(sprintf "Memory usage: %.2f MB" (float delta))

    [<Fact(Skip = SKIP)>]
    let ``Profile`` () =
        let sut = ConcurrentTrie<int>()

        let add _ =
            for _ = 1 to 10000 do
                sut.Add(Guid.NewGuid().ToString(), 0)

        profile <| fun () -> Array.Parallel.iter add [| 1..1000 |]

    [<Fact(Skip = SKIP)>]
    let ``Profile add/remove`` () =
        let sut = ConcurrentTrie<int>()

        let key(i, j) = $"{i}-{j}"

        let addRemove j =
            for i in 1..10000 do
                sut.Add(key(i, j), i)
            for i in 1..10000 do
                sut.Remove(key(i, j)) |> ignore

        profile <| fun () -> Array.Parallel.iter addRemove [| 1..1000 |]
